### PR TITLE
Added L1 regularization

### PR DESCRIPTION
**Pull request by eladhoffer**

### CODE REVIEW

File: sgd.lua [hunk 1]
```diff
@@ -53,6 +55,13 @@ function optim.sgd(opfunc, x, config, state)
       dfdx:add(state.decayParameters)
    end
 
+   -- (2.1) L1 regularization
+   if L1coeff > 0 then
+       dfdx:add(L1coeff, torch.sign(x))
```
Comment (soumith) [inline sgd.lua]: torch.sign(x) allocates a new result tensor at every call. This is bad practice, you ideally want to reuse the same buffer every time (see line 51 to 55 in your version).

**Comment by soumith** [comment]
apart from the in-line comment, i would suggest more consistent naming.
optim.sdg already has .weightDecay. For L1 decay, you could name it .L1WeightDecay, or .weightDecayL1 which is more in line with the current option (i am also open to better naming).

**Comment by pakozm** [comment]
I apologize in case this comment is out of scope for this package in Torch 7. I'm not an expert in Torch, but I'm aware of L1 penalty term and its implementation issues. This comment is just a note, which can be taken into account or not, feel free! ;-) 
The proposed approach can lead to unstable weights instead of real zero weigths, that is, the weights can be cycling from positive to negative values from one iteration to another. There are simple ways to solve this issue, normally applying some kind of "truncated gradient" solution, as the one used in [scikit-learn](http://scikit-learn.org/stable/modules/sgd.html#implementation-details). I have my own implementation of L1 (not for Torch) and I found very useful the truncation of weights just to ensure they are set to zero.
In the most simple approach, this truncation is applied by:

1. Computing *which* weights will cross zero after applying the L1 penalty
2. Applying the proposed update to all weights
3. Explicitly fill with zero the weights annotated as zero in step (1)

Obviously different approaches are possible for truncated gradient.

**Comment by koraykv** [comment]
I am going to close this request since it hasn't moved much.

**Pull request closed without merging**